source/adapters/level_zero/device.cpp

```diff
@@ -69,8 +69,16 @@ UR_APIEXPORT ur_result_t UR_APICALL urDeviceGet(
       urPrint("Unknown device type");
       break;
     }
-    if (Matched)
+    // Filter out composite devices when ZE_FLAT_DEVICE_HIERARCHY=COMBINED:
+    // component devices (tiles) enable ZE_DEVICE_PROPERTY_FLAG_SUBDEVICE while
+    // composite devices (cards) don't.
+    bool isComposite =
+        (D->ZeDeviceProperties->flags & ZE_DEVICE_PROPERTY_FLAG_SUBDEVICE) == 0;
+    const char *Mode = std::getenv("ZE_FLAT_DEVICE_HIERARCHY");
```

Comment:
`ur_getenv`

But a broader question, should UR read environment variables defined by other components? The environment can change between two getenv calls, resulting in potential subtle bugs.

Reply:
I understand the concern, but as far as I know, this envvar should never change once the application has started. I'm open to suggestions here, but I don't see a better way to deal with this.

Reply:
> envvar should never change once the application has started

```
man 3 getenv

The setenv() function adds the variable name to the environment with the value value .... 
If name does exist in the environment, then its value is changed to value if overwrite is nonzero.
```

```
man 3 putenv

The putenv() function adds or changes the value of environment variables.
The argument string is of the form name=value. ...
If name does exist, then the value of name in the environment is changed to value.
```

The environment change could be unintentional. This is a TOCTOU-like issue.

> I'm open to suggestions here

Ideally L0 would expose what is the device hierarchy it was configured with. Alternatively, maybe it would be possible to infer the hierarchy based on the behavior of `zeGetDevice`/`zeDeviceGetSubDevice` as defined [here](https://spec.oneapi.io/level-zero/latest/core/PROG.html#device-hierarchy)?

Reply:
@pbalcer I have just pushed a commit that removes all readings of `ZE_FLAT_DEVICE_HIERARCHY`. Let me know if that addresses your concerns.